src/bun.js/bindings/JSStringDecoder.cpp

```diff
@@ -0,0 +1,383 @@
+}
+
+// Checks at most 3 bytes at the end of a Buffer in order to detect an
+// incomplete multi-byte UTF-8 character. The total number of bytes (2, 3, or 4)
+// needed to complete the UTF-8 character (if applicable) are returned.
+uint8_t JSStringDecoder::utf8CheckIncomplete(JSC::JSUint8Array* buf, uint32_t i)
+{
+    uint32_t j = buf->length() - 1;
+    uint8_t* bufPtr = buf->typedVector();
+    if (j < i) return 0;
+    int8_t nb = utf8CheckByte(bufPtr[j]);
+    if (nb >= 0) {
+        if (nb > 0) m_lastNeed = nb - 1;
+        return nb;
+    }
+    if (--j < i || nb == -2) return 0;
+    nb = utf8CheckByte(bufPtr[j]);
+    if (nb >= 0) {
+        if (nb > 0) m_lastNeed = nb - 2;
+        return nb;
+    }
+    if (--j < i || nb == -2) return 0;
+    nb = utf8CheckByte(bufPtr[j]);
+    if (nb >= 0) {
+        if (nb > 0) {
+            if (nb == 2) nb = 0;else m_lastNeed = nb - 3;
+        }
+        return nb;
+    }
+    return 0;
+}
+
+JSC::JSValue JSStringDecoder::text(JSC::VM& vm, JSC::JSGlobalObject* globalObject, JSC::JSUint8Array* buf, uint32_t offset)
+{
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+    uint32_t length = buf->length();
+    uint8_t* bufPtr = buf->typedVector();
+
+    switch (m_encoding) {
+    case BufferEncodingType::ucs2:
+    case BufferEncodingType::utf16le: {
+        if ((length - offset) % 2 == 0) {
+            JSString* r = JSC::JSValue::decode(Bun__encoding__toString(bufPtr + offset, length - offset, globalObject, static_cast<uint8_t>(m_encoding))).toString(globalObject);
+            RETURN_IF_EXCEPTION(throwScope, JSC::jsUndefined());
+            if (r->length()) {
+                UChar c = r->value(globalObject).characterAt(r->length() - 1);
+                if (c >= 0xD800 && c <= 0xDBFF) {
+                    m_lastNeed = 2;
+                    m_lastTotal = 4;
+                    m_lastChar[0] = bufPtr[length - 2];
+                    m_lastChar[1] = bufPtr[length - 1];
+                    return JSC::jsSubstring(vm, globalObject, r, 0, r->length() - 1);
+                }
+            }
+            return r;
+        }
+        m_lastNeed = 1;
+        m_lastTotal = 2;
+        m_lastChar[0] = bufPtr[length - 1];
+        return JSC::JSValue::decode(Bun__encoding__toString(bufPtr + offset, length - offset - 1, globalObject, static_cast<uint8_t>(m_encoding)));
+    }
+    case BufferEncodingType::utf8: {
+        uint32_t total = utf8CheckIncomplete(buf, offset);
+        if (!m_lastNeed)
+            return JSC::JSValue::decode(Bun__encoding__toString(bufPtr + offset, length - offset, globalObject, static_cast<uint8_t>(m_encoding)));
+        m_lastTotal = total;
+        uint32_t end = length - (total - m_lastNeed);
+        if (end < length)
+            memmove(m_lastChar, bufPtr + end, std::min(4U, length - end));
+        return JSC::JSValue::decode(Bun__encoding__toString(bufPtr + offset, end - offset, globalObject, static_cast<uint8_t>(m_encoding)));
```

Comment:
All the returns here should be `RELEASE_AND_RETURN(throwScope, JSC::SJValue ...

Reply:
fixed.